gtk/src/toga_gtk/widgets/base.py

```diff
@@ -119,33 +117,40 @@ def apply_css(self, property, css, native=None, selector=".toga"):
         if native is None:
             native = self.native
 
-        style_context = native.get_style_context()
         style_provider = self.style_providers.pop((property, id(native)), None)
 
         # If there was a previous style provider for the given property, remove
         # it from the GTK widget
         if style_provider:
-            style_context.remove_provider(style_provider)
-
-        # If there's new CSS to apply, construct a new provider, and install it.
-        if css is not None:
-            # Create a new CSS StyleProvider
+            Gtk.StyleContext.remove_provider_for_display(
+                Gdk.Display.get_default(),
+                style_provider,
+            )
+        # If there's new CSS to apply, install it.
+        if css:
             style_provider = Gtk.CssProvider()
             styles = " ".join(f"{key}: {value};" for key, value in css.items())
-            declaration = selector + " {" + styles + "}"
-            style_provider.load_from_data(declaration.encode())
-
-            # Add the provider to the widget
-            style_context.add_provider(
+            declaration = selector + f"#{native.get_name()}" + " {" + styles + "}"
+
+            # Backward compatibility fix for different gtk versions ===========
+            if Gtk.get_major_version() >= 4 and Gtk.get_minor_version() >= 12:
+                style_provider.load_from_string(declaration)
+            elif Gtk.get_major_version() >= 4 and Gtk.get_minor_version() > 8:
+                style_provider.load_from_data(declaration, len(declaration))
+            else:
+                style_provider.load_from_data(declaration.encode("utf-8"))
+            # =================================================================
```

Comment:
> Is there any way to reduce the number of code branches here without destroying the user experience?

I tried to find other way to overcome this problem and the only possible way is to write our css to file and read this file using `load_from_file` which is consistence on all version as I know. 

> what versions of GTK are we actually going to see in the wild

On Debian 12 the GTK version is 4.8.x. As the other version is still mature we can count on this version.

Reply:
So - the more I look into this, the more I'm having questions about how this PR can be successfully landed, because GTK4 support seems like something that isn't widespread enough to actually support being used. I'll put my thoughts in a comment on the main PR, rather than burying them on this specific change.